go/fixchain/logger.go

```diff
@@ -0,0 +1,232 @@
+	// active servers, because we can't close it to signal the
+	// end, because of retries.
+	wg      sync.WaitGroup
+	limiter Limiter
+
+	postCertCache  *lockedMap
+	postChainCache *lockedMap
+}
+
+// IsPosted tells the caller whether a chain for the given certificate has
+// already been successfully posted to the log by this Logger.
+func (l *Logger) IsPosted(cert *x509.Certificate) bool {
+	return l.postCertCache.get(hash(cert))
+}
+
+// QueueChain adds the given chain to the queue to be posted to the log.
+func (l *Logger) QueueChain(chain []*x509.Certificate) {
+	if chain == nil {
+		return
+	}
+
+	l.posted++
+	// Has a chain for the cert this chain if for already been successfully
+	//posted to the log by this Logger?
+	h := hash(chain[0]) // Chains are cert -> root
+	if l.postCertCache.get(h) {
+		l.reposted++
+		return // Don't post chain for a cert that has already had a chain posted.
+	}
+	// If we assume all chains for the same cert are equally
+	// likely to succeed, then we could mark the cert as posted
+	// here. However, bugs might cause a log to refuse one chain
+	// and accept another, so try each unique chain.
+
+	// Has this Logger already tried to post this chain?
+	h = hashChain(chain)
+	if l.postChainCache.get(h) {
+		l.chainReposted++
+		return
+	}
+	l.postChainCache.set(h, true)
+
+	p := &toPost{chain: chain, retries: 5}
+	l.postToLog(p)
+}
+
+// Wait for all of the active requests to finish being processed.
+func (l *Logger) Wait() {
+	l.wg.Wait()
+}
+
+// RootCerts returns the root certificates that the log accepts.
+func (l *Logger) RootCerts() *x509.CertPool {
+	if l.roots == nil {
+		l.roots = l.getRoots()
+	}
+	return l.roots
+}
+
+func (l *Logger) getRoots() *x509.CertPool {
+	rootsJSON, err := l.client.Get(l.url + "/ct/v1/get-roots")
+	if err != nil {
+		log.Fatalf("can't get roots from %s: %s", l.url, err)
+	}
+	defer rootsJSON.Body.Close()
+	if rootsJSON.StatusCode != 200 {
+		log.Fatalf("can't deal with status other than 200: %d", rootsJSON.StatusCode)
+	}
+	j, err := ioutil.ReadAll(rootsJSON.Body)
+	if err != nil {
+		log.Fatalf("can't read roots: %s", err)
+	}
+	type Certificates struct {
+		Certificates [][]byte
+	}
+	var certs Certificates
+	err = json.Unmarshal(j, &certs)
+	if err != nil {
+		log.Fatalf("can't parse json (%s): %s", err, j)
+	}
+	ret := x509.NewCertPool()
+	for i := 0; i < len(certs.Certificates); i++ {
+		r, err := x509.ParseCertificate(certs.Certificates[i])
+		switch err.(type) {
+		case nil, x509.NonFatalErrors:
+			// ignore
+		default:
+			log.Fatalf("can't parse certificate: %s %#v", err, certs.Certificates[i])
+		}
+		ret.AddCert(r)
+		log.Printf("Root %d: %s", i, r.Subject.CommonName)
+	}
+	return ret
+}
+
+type toPost struct {
+	chain   []*x509.Certificate
+	retries uint8
+}
+
+func (l *Logger) postToLog(p *toPost) {
+	// Add post to toPost channel in a separate goroutine to avoid deadlock
+	// between postServers during retries.
+	l.wg.Add(2)
```

Comment:
After thinking this through, I have changed this back to Add(1) and removed the Done() inside the goroutine.